File: Cloud Sheets Samples/Office365 Excel/c++/office365excel.cpp
/*
 * Cloud Sheets 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Sheets in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudsheets
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <sstream>
#include <cctype>
#include "../../include/cloudsheets.h"

class Office365Demo {
private:
    bool running;
    char* current_workbook_id;
    bool isWorkbook = false;
    std::string promptPrefix = "Excel";
    Office365Excel excel;

    // Utility methods
    static std::string trim(const std::string& str) {
        auto start = str.begin();
        while (start != str.end() && std::isspace(*start)) {
            start++;
        }

        auto end = str.end();
        do {
            end--;
        } while (std::distance(str.begin(), end) > 0 && std::isspace(*end));

        return std::string(start, end + 1);
    }

    static std::string to_lowercase(std::string str) {
        std::transform(str.begin(), str.end(), str.begin(),
            [](unsigned char c) { return std::tolower(c); });
        return str;
    }
    bool isEmpty(const char* str) const {
        return (str == nullptr || str[0] == '\0');
    }

    // Parse input into command and arguments
    std::vector<std::string> parse_input(const std::string& input) {
        std::vector<std::string> tokens;
        std::istringstream iss(trim(input));
        std::string token;

        while (iss >> token) {
            tokens.push_back(token);
        }

        return tokens;
    }

    void authenticate(const char* client, const char* secret)
    {
        OAuth oauth;
        const char* oAuthScope = "offline_access files.readwrite user.read";
        const char* oAuthAuthUrl = "https://login.microsoftonline.com/common/oauth2/v2.0/authorize";
        const char* oAuthTokenUrl = "https://login.microsoftonline.com/common/oauth2/v2.0/token";
        oauth.SetClientId(client);
        oauth.SetClientSecret(secret);
        oauth.SetAuthorizationScope(oAuthScope);
        oauth.SetServerAuthURL(oAuthAuthUrl);
        oauth.SetServerTokenURL(oAuthTokenUrl);
        excel.SetAuthorization(oauth.GetAuthorization());

    }

    // Command implementations
    void listWorkbooks() {
        excel.ListWorkbooks();
        const int bookCount = excel.GetWorkbooks()->GetCount();

        std::cout << std::string(20, '-') << std::endl;
        for (int i = 0; i < bookCount; i++)
        {
            std::cout << excel.GetWorkbooks()->Get(i)->GetName() << std::endl;
        }
    }

    char* getWorkbookIdByName(const std::string& workbook_name)
    {
        excel.ListWorkbooks();
        const int bookCount = excel.GetWorkbooks()->GetCount();

        for (int i = 0; i < bookCount; i++)
        {
            if (excel.GetWorkbooks()->Get(i)->GetName() == workbook_name)
            {
                return excel.GetWorkbookId(i);
           }
        }
        return nullptr;
    }

    void displayMenu() {
        if (!isWorkbook) {
            std::cout << "h                 -  Help\n";
            std::cout << "ls                -  List workbooks\n";
            std::cout << "select [workbook] -  Select a workbook\n";
            std::cout << "new    [name]     -  Create new Workbook\n";
            std::cout << "delete [name]     -  Delete Workbook\n";
            std::cout << "q                 -  Quit\n";
         }
        else {
            std::cout << "ls                - List sheets\n";
            std::cout << "add    [sheet]    - Add new sheet\n";
            std::cout << "update [sheet]    - Update sheet\n";
            std::cout << "delete [sheet]    - Delete sheet\n";
            std::cout << "exit              - Return to Main Menu\n";
        }
    }

    void selectWorkbook(const std::string& workbook_name) {
        current_workbook_id = getWorkbookIdByName(workbook_name);
        if (isEmpty(current_workbook_id))
        {
            std::cout << "Workbook not found, please try again." << std::endl;
            return;
        }
        excel.LoadWorkbook(current_workbook_id);
        isWorkbook = true; 
        promptPrefix = workbook_name;
        std::cout << "Selected workbook: " << workbook_name << std::endl;
        displayMenu();
    }

    void createWorkbook(const std::string& workbook_name) {

        excel.CreateWorkbook(workbook_name.c_str());
        std::cout << "Created workbook: " << workbook_name << std::endl;
    }

    void addSheet(const std::string& sheet_name) {
        excel.AddWorksheet(sheet_name.c_str());
        std::cout << "Added sheet: " << sheet_name << std::endl;
        
    }
    void listSheets()
    {
        for (int i = 0; i < excel.GetWorksheets()->GetCount(); i++)
        {
            std::cout << excel.GetWorksheets()->Get(i)->GetName() << std::endl;
        }                 
    }
    const char* getSheetIdByName(const std::string& sheetName)
    {
        if (isWorkbook)
        {
            for (int i = 0; i < excel.GetWorksheets()->GetCount(); i++)
            {
                auto* found = excel.GetWorksheets()->Get(i)->GetName();
                if (excel.GetWorksheets()->Get(i)->GetName() == sheetName)
                {
                    return excel.GetWorksheets()->Get(i)->GetWorksheetId();
                }
            }
            return nullptr;
        }
    }

    void updateWorkbook(const std::string& workbook_id) {
        std::cout << "Updating workbook: " << workbook_id << std::endl;
    }

    std::string normalizeArg(const std::string& arg) {
        if (!arg.empty()) {
            // Check if the string already ends with .xlsx
            if (arg.size() >= 5 &&
                arg.compare(arg.size() - 5, 5, ".xlsx") == 0) {
                return arg;
            }
            // Append .xlsx if not already present
            return arg + ".xlsx";
        }
        // Return empty string for empty input 
        return "";
    }

    void updateWorksheet(const std::string& worksheet_name) {

        if (isWorkbook)
        {
            const char* sheetId = getSheetIdByName(worksheet_name.c_str());
            if (sheetId == nullptr)
            {
                std::cout << "The selected sheet was not found in the workbook\n";
                return;
            }
            std::string newName;
            std::cout << "Enter new sheet name: ";
            std::getline(std::cin, newName); 
            excel.LoadWorksheet(sheetId);
            excel.GetWorksheets()->Get(0)->SetName(newName.c_str());
            excel.UpdateWorksheet();
            std::cout << "Sheet updated successfully! " << std::endl;

        }
    }

    void deleteWorkbook(const std::string& workbook_name) {

        char* workbook_id = getWorkbookIdByName(workbook_name);
        if (!isEmpty(workbook_id))
        {
            
            excel.DeleteWorkbook(workbook_id);
            if (current_workbook_id == workbook_id) {
                current_workbook_id[0] = '\n';
            }
            std::cout << "Workbook " << workbook_name <<" was deleted successfully!" << std::endl;
            return;
        }

        std::cout << "Workbook " << workbook_name <<" was not found!" << std::endl;
    }

    void deleteSheet(const std::string& sheet_name) {
        const char* sheetId = getSheetIdByName(sheet_name);
        if (sheetId ==nullptr)
        {
            std::cout << "The selected sheet was not found in the workbook";
            return;
        }
        excel.LoadWorksheet(sheetId);
        excel.DeleteWorksheet(sheetId);
        std::cout << "Deleted sheet: " << sheet_name << std::endl;
    }

    // Process command
    void processCommand(const std::string& input) {
        if (input.empty()) return;
        auto args = parse_input(input);

        if (args.empty()) return;

        std::string command = to_lowercase(args[0]);

        if (command == "ls") {
            if (isWorkbook)
            {
                listSheets();
                return;
            }
            listWorkbooks();
        }
        else if (command == "?" || command == "h") {
            displayMenu();
        }
        else if (command == "select" && args.size() > 1) {
            selectWorkbook(normalizeArg(args[1]));
        }
        else if (command == "new") {
            if (args.size() < 2) {
                std::cout << "Please provide a workbook name.\n";
                return;
            }
            createWorkbook(args[1]);
        }
        else if (command == "add") {
            if (!isWorkbook)
            {
                std::cout << "Please select a Workbook first \n";
                return;
            }
            if (args.size() < 2) {
                std::cout << "Please provide a sheet name.\n";
                return;
            }
            addSheet(args[1]);
        }
        else if (command == "us" && args.size() > 1) {
            updateWorkbook(args[1]);
        }
        else if (command == "update" && args.size() > 1) {
            updateWorksheet(args[1]);
        }
        else if ((command == "del" || command == "delete") && args.size() > 1) {
            if (!isWorkbook) {
                deleteWorkbook(normalizeArg(args[1]));
                return;
            }
            else {
                deleteSheet(args[1]);
            }
        }
        else if (command == "exit" || command == "q") {
            if (!isWorkbook)
            {
                std::cout << "Goodbye!\n";
                running = false;
            }
            isWorkbook = false;
            promptPrefix = "Excel";

        }
        else {
            std::cout << "Invalid command. Please try again.\n";
        }
    }

public:
    Office365Demo() : running(true) {}

    void run(const char* client, const char* secret) {
        excel.CreateSession(true);

        displayMenu();

        authenticate(client,secret);
        while (running) {
            std::cout << promptPrefix + ">";
            std::string input;
            std::getline(std::cin, input);

            try {
                processCommand(input);
            }
            catch (const std::exception& e) {
                std::cerr << e.what() << std::endl;
            }
        }
    }
};

int main() {
    Office365Demo excel;

    std::string clientId;
    std::string clientSecret;
    std::cout << "Enter the Id of the client assigned when registering the application: ";
    std::getline(std::cin, clientId);
    std::cout << "Enter the secret value for the client assigned when registering the application: ";
    std::getline(std::cin, clientSecret);

    excel.run(clientId.c_str(), clientSecret.c_str());
    return 0;
}

File: Cloud Sheets Samples/Office365 Excel/javascript/office365excel.js
/*
 * Cloud Sheets 2024 JavaScript Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Sheets in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudsheets
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
 
const readline = require("readline");
const cloudsheets = require("@nsoftware/cloudsheets");

if(!cloudsheets) {
  console.error("Cannot find cloudsheets.");
  process.exit(1);
}
let rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});


class Office365ExcelDemo {
    constructor() {
        this.oAuthScope = "offline_access files.readwrite user.read";
        this.oAuthAuthUrl = "https://login.microsoftonline.com/common/oauth2/v2.0/authorize";
        this.oAuthTokenUrl = "https://login.microsoftonline.com/common/oauth2/v2.0/token";
        this.excel = null;
        this.currentWorkbookId = "";
        this.selectedWorkbookName="";
        this.currentSheetId = -1;
        this.changesMade = false;
        this.currentSheet = null;
        this.rl = rl;

        this.running = true;
    }

    askQuestion(query) {
        return new Promise(resolve => {
            this.rl.question(query, resolve);
        });
    }
    reset = async () => 
    {
        this.currentWorkbookId="";
        await this.excel.listWorkbooks();
    }
    normalizeArg = (arg) => 
    {
        if(arg)
        {
            return `${arg}${arg.endsWith('.xlsx') ? '' : '.xlsx'}`;
        }
    }
    getNumberInput= async (prompt) => {
        const input = (await this.askQuestion(prompt)).trim();
        const num = parseInt(input, 10);
        return (isNaN(num) || num < 0) ? null : num;
    }

    getSheetId = async(sheetName) => 
    {
        if(!sheetName)
        {
            sheetName = await this.askQuestion("Sheet Name:")
        }
        const worksheets = this.excel.getWorksheets();
        for (let i = 0; i < worksheets.size(); i++) {
            const sheet = worksheets.get(i);
            if (sheet.getName() === sheetName) return sheet.getWorksheetId();
        }
        return null; // Not found
    }
    
    getConfirmation = async(question) =>
    {
        const answer = await this.askQuestion(question);
        return answer.toLowerCase() === 'y';
    }
    getCommand = async() => 
    {
        const prompt = this.currentWorkbookId 
        ? `${this.selectedWorkbookName}>`
        : 'Excel> ';
        return await this.askQuestion(prompt);
    }
    
    selectWorkbook = async(WorkbookName) => 
    {
        if (!WorkbookName.endsWith('.xlsx'))
        {
            WorkbookName+='.xlsx'
        }
        try {
            await this.loadWorkbook(WorkbookName);
            this.displayWorksheetMenu();
            
        }catch(error)
        {
            console.error("Error selecting Workbook:", error);
        }
    }
    loadWorkbook = async(WorkbookName) =>
    {
        if(!WorkbookName)
        {
            WorkbookName = await this.askQuestion("Enter the Workbook Name: ");
        }
        this.currentWorkbookId = await this.getWorkbookId(WorkbookName);
        if(this.currentWorkbookId != null)
        {
            await this.excel.loadWorkbook(this.currentWorkbookId);
            this.selectedWorkbookName = WorkbookName
        }
        else
        {
            await this.loadWorkbook()
        }
    }
    displayWorksheetMenu() {
        if (!this.selectedWorkbookName ==="") {
            console.log("No Worksheet selected.");
            return;
        }
        console.log(
            "\nls\t-\tList Worksheets\n" +
            "add\t-\tAdd Worksheet\n" +
            "update\t-\tUpdate Worksheet\n" +
            "delete\t-\tDelete Worksheet\n" +
            "exit\t-\tReturn to Main Menu\n"
        );
    }
    displayMainMenu() {
        console.log(
                "?\t-\tHelp\n" +
                "ls\t-\tList Workbook\n" +
                "select\t-\tSelect Workbook\n" +
                "new\t-\tCreate Workbook\n" +
                "del\t-\tDelete Workbook\n" +
                "q\t-\tQuit\n"
        );
    }
    listSheets = async() => 
    {
        const sheets = this.excel.getWorksheets()
        for (let i = 0; i < sheets.size(); i++) {
            console.log("Title: " + sheets.get(i).getName());
        }
    }

    getWorkbookId = async (WorkbookName) =>
    {
        const Workbooks = this.excel.getWorkbooks();
        for (let i = 0; i < Workbooks.size(); i++) {
            const name = Workbooks.get(i).getName();
            if(name === WorkbookName)
            {
                return Workbooks.get(i).getId();
            }
        }
        if (!this.currentWorkbookId) {
            console.log(`No Workbook found with name: ${WorkbookName}`);
            return null;
        }
    };
    listWorkbooks = async () => {
        if(this.currentWorkbookId != "")
        {
            await this.listSheets();
            return;
        }
        await this.excel.listWorkbooks();
        const Workbooks = this.excel.getWorkbooks();
        const maxNameLength = 24;

        console.log("Name".padEnd(maxNameLength));
        console.log("-".repeat(20)); 

        for (let i = 0; i < Workbooks.size(); i++) {
            const name = Workbooks.get(i).getName();
            console.log(`${name}`);
        }
    };
    createWorkbook = async(WorkbookName) => 
    {
        if(!WorkbookName)
        {
            WorkbookName = await this.askQuestion("Enter Workbook name: ");
        }
        await this.excel.createWorkbook(WorkbookName);
        console.log(`Workbook ${WorkbookName} created successfully.`)
        this.reset()
    };
    addSheet = async(sheetTitle) => 
    {
        if(this.selectedWorkbookName === "")
        {
            console.log("A Workbook must first be selected")
            return;
        }
        if(!sheetTitle)
        {
            sheetTitle = await this.askQuestion("Enter Sheet title: ");
        }
        await this.excel.addWorksheet(sheetTitle);
        console.log("Sheet added successfully!")
    };

    updateWorksheet = async(sheetTitle) => 
    {
        if(this.selectedWorkbookName === "")
        {
            console.log("A Workbook must first be selected")
            return;
        }
        if(!sheetTitle)
        {
            sheetTitle = await this.askQuestion("Enter Sheet title: ");
        }

        const sheetId = await this.getSheetId(sheetTitle);

        await this.excel.loadWorksheet(sheetId); 
        if(await this.getConfirmation("Change sheet name (Y/N)? "))
        {
            const newName = await this.askQuestion("Enter new sheet name: ")
            this.excel.getWorksheets().get(0).setName(newName);
            await this.excel.updateWorksheet();
        }
    };

    deleteSheet = async(sheetTitle) => 
    {
        if(this.selectedWorkbookName === "")
        {
            console.log("A Workbook must first be selected")
            return;
        }
        if(!sheetTitle)
        {
            sheetTitle = await this.askQuestion("Enter Sheet title: ");
        }
        const sheetId = await this.getSheetId(sheetTitle);

        await this.excel.loadWorksheet(sheetId);
        if(await this.getConfirmation("Delete Sheet (Y/N)? "))
        {
            await this.excel.deleteWorksheet(sheetId);
        }
    }
    deleteWorkbook = async(WorkbookName) => 
    {  
        await this.excel.listWorkbooks();
        await this.loadWorkbook(WorkbookName);
        await this.excel.deleteWorkbook(this.currentWorkbookId)
        await this.reset()
    }

    exit = async()=>
    {
        this.currentWorkbookId ="";
        this.currentSheetId="";
        await this.excel.listWorkbooks();
    }
    async main(args) {
        if (args.length < 4) {
            console.log("usage: node googlesheets.js -c client -s secret\n" +
                " -c The Id of the client assigned when registering the application.\n" +
                " -s The secret value for the client assigned when registering the application.");
            return;
        }

        try {
            const myArgs = this.parseArgs(args);
            this.excel = new cloudsheets.office365excel();


            await this.authenticate(myArgs.get("c"), myArgs.get("s"));
            console.log("Loading Workbooks...");
            await this.excel.listWorkbooks();
            this.currentWorkbookId ? this.displayWorksheetMenu() : this.displayMainMenu();

            while (this.running) {
                try {
                    const input= await this.getCommand();
                    const [command, ...args] = input.trim().split(/\s+/);

                    // Process commands
                    switch (command.toLowerCase().trim()) {
                        case 'ls':
                            await this.listWorkbooks();
                            break;
                        case '?':
                            this.currentWorkbookId ? this.displayWorksheetMenu() : this.displayMainMenu();
                            break;
                        case 'select':
                            await this.selectWorkbook(this.normalizeArg(args[0]));
                            break;
                        case 'new':
                            await this.createWorkbook(args[0]);
                            break;
                        case 'add': 
                            await this.addSheet(args[0]);
                            break;
                        case 'us': 
                            await this.updateWorkbook(args[0]);
                            break;
                        case 'update': 
                            await this.updateWorksheet(args[0]);
                            break;
                        case 'del': 
                            await this.deleteWorkbook(this.normalizeArg(args[0]));
                            break;
                        case 'delete': 
                            await this.deleteSheet(args[0]);
                            break;
                        case 'exit': 
                            await this.exit();
                            break;
                        case 'q':
                            console.log('Goodbye!');
                            this.rl.close();
                            this.running = false;
                            return;
                        default:
                            console.log('Invalid command. Please try again.');
                    }
                } catch (error) {
                    console.error('An error occurred:', error.message);
                    break;
                }
            }

        } catch (ex) {
            console.log(ex.message);
        }

    }

    async authenticate(client, secret) {
        if (client && secret) {
            const oauth = new cloudsheets.oauth();
            oauth.setClientId(client)
            oauth.setClientSecret(secret);
            oauth.setAuthorizationScope(this.oAuthScope);
            oauth.setServerAuthURL(this.oAuthAuthUrl);
            oauth.setServerTokenURL(this.oAuthTokenUrl);

            try {
                let authorization = await oauth.getAuthorization()
                await this.excel.setAuthorization(authorization);
            } catch (ex) {
                console.log(ex.message);
            }
        } else {
            console.log("Client Id and Secret cannot be null!");
        }
    }
    parseArgs(args) {
        const myArgs = new Map();
        for (let i = 0; i < args.length; i += 2) {
            if (args[i].startsWith('-')) {
                myArgs.set(args[i].substring(1), args[i + 1]);
            }
        }
        return myArgs;
    }
}

// Run the demo
const demo = new Office365ExcelDemo();
demo.main(process.argv.slice(2));

function prompt(promptName, label, punctuation, defaultVal)
{
  lastPrompt = promptName;
  lastDefault = defaultVal;
  process.stdout.write(`${label} [${defaultVal}]${punctuation} `);
}
